File: packages/tolk/test/parity/cast_f16_to_f32/main.ml
(* Parity case: float16 to float32 cast. *)

open Tolk_uop
module B = Program_spec_builder

let global_ptr dt = Dtype.Ptr.create dt ~addrspace:Global ~size:(-1)

let kernel () =
  let from_dt = Dtype.Val.float16 in
  let to_dt = Dtype.Val.float32 in
  let from_ptr = global_ptr from_dt in
  let to_ptr = global_ptr to_dt in
  let b = B.create () in
  let p0 = B.emit b (Param { slot = 0; dtype = from_ptr }) in
  let p1 = B.emit b (Param { slot = 1; dtype = to_ptr }) in
  let c0 =
    B.emit b
      (Const { value = Const.int Dtype.Val.int32 0; dtype = Dtype.Val.int32 })
  in
  let idx0 =
    B.emit b
      (Index { ptr = p0; idxs = [ c0 ]; dtype = from_ptr })
  in
  let idx1 =
    B.emit b (Index { ptr = p1; idxs = [ c0 ]; dtype = to_ptr })
  in
  let ld = B.emit b (Load { src = idx0; alt = None; gate = None; dtype = from_dt }) in
  let cast = B.emit b (Cast { src = ld; dtype = to_dt }) in
  let _ = B.emit b (Store { dst = idx1; value = cast; gate = None }) in
  B.finish b

let () = Helpers.dump_stage7_program ~out_dir:Sys.argv.(1) (kernel ())

File: packages/tolk/test/parity/cast_f16_to_f32/main.py
#!/usr/bin/env python3
"""Parity case: float16 to float32 cast."""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from helpers import dump_stage7_program  # noqa: E402

from tinygrad.uop.ops import UOp, Ops, KernelInfo  # noqa: E402
from tinygrad.dtype import dtypes  # noqa: E402


def kernel():
    sink = UOp(Ops.SINK, dtypes.void, (), arg=KernelInfo())
    a = UOp.param(0, dtypes.half, shape=(-1,))
    b = UOp.param(1, dtypes.float32, shape=(-1,))
    idx = UOp.const(dtypes.int, 0)
    idx_a = a.index(idx)
    ld = UOp(Ops.LOAD, dtypes.half, (idx_a,))
    cast = UOp(Ops.CAST, dtypes.float32, (ld,))
    idx_b = b.index(idx)
    store = UOp(Ops.STORE, dtypes.void, (idx_b, cast))
    return [sink, a, b, idx, idx_a, ld, cast, idx_b, store]


if __name__ == "__main__":
    dump_stage7_program(kernel(), os.path.dirname(os.path.abspath(__file__)))
